storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java

```diff
@@ -0,0 +1,749 @@
+        }
+        return false;
+    }
+    
+    static DynamicState stateMachineStep(DynamicState dynamicState, StaticState staticState) throws Exception {
+        LOG.debug("STATE {}", dynamicState.state);
+        switch (dynamicState.state) {
+            case EMPTY:
+                return handleEmpty(dynamicState, staticState);
+            case RUNNING:
+                return handleRunning(dynamicState, staticState);
+            case WATING_FOR_WORKER_START:
+                return handleWaitingForWorkerStart(dynamicState, staticState);
+            case KILL_AND_RELAUNCH:
+                return handleKillAndRelaunch(dynamicState, staticState);
+            case KILL:
+                return handleKill(dynamicState, staticState);
+            case WAITING_FOR_BASIC_LOCALIZATION:
+                return handleWaitingForBasicLocalization(dynamicState, staticState);
+            case WAITING_FOR_BLOB_LOCALIZATION:
+                return handleWaitingForBlobLocalization(dynamicState, staticState);
+            default:
+                throw new IllegalStateException("Code not ready to handle a state of "+dynamicState.state);
+        }
+    }
+    
+    /**
+     * Prepare for a new assignment my downloading new required blobs, or going to empty if there is nothing to download.
+     * PRECONDITION: The slot should be empty
+     * @param dynamicState current state
+     * @param staticState static data
+     * @return the next state
+     */
+    static DynamicState prepareForNewAssignmentOnEmptySlot(DynamicState dynamicState, StaticState staticState) {
+        assert(dynamicState.container == null);
+        
+        if (dynamicState.newAssignment == null) {
+            return dynamicState.withState(MachineState.EMPTY);
+        }
+        Future<Void> pendingDownload = staticState.localizer.requestDownloadBaseTopologyBlobs(dynamicState.newAssignment.get_topology_id(), staticState.port);
+        return dynamicState.withPendingLocalization(dynamicState.newAssignment, pendingDownload).withState(MachineState.WAITING_FOR_BASIC_LOCALIZATION);
+    }
+    
+    /**
+     * Kill the current container and start downloading what the new assignment needs, if there is a new assignment
+     * PRECONDITION: container != null
+     * @param dynamicState current state
+     * @param staticState static data
+     * @return the next state
+     * @throws Exception 
+     */
+    static DynamicState killContainerForChangedAssignment(DynamicState dynamicState, StaticState staticState) throws Exception {
+        assert(dynamicState.container != null);
+        
+        staticState.iSupervisor.killedWorker(staticState.port);
+        dynamicState.container.kill();
+        Future<Void> pendingDownload = null;
+        if (dynamicState.newAssignment != null) {
+            pendingDownload = staticState.localizer.requestDownloadBaseTopologyBlobs(dynamicState.newAssignment.get_topology_id(), staticState.port);
+        }
+        Time.sleep(1000);
```

Comment:
You are right that I missed this.

https://github.com/apache/storm/blob/master/storm-core/src/jvm/org/apache/storm/daemon/supervisor/workermanager/DefaultWorkerManager.java#L268

I'll fix it.